Add extra generation() configuration knobs to run_generation.py

vivekgoe opened the pull request:
# What does this PR do?
This PR exposes generation extra configuration knobs such as "bad_words_ids", "force_words_ids", "num_return_sequences" to user of run_generation.py.  Also adds code to support constrained_beam_search on HPU (must to support force_words_ids).

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

regisss commented on examples/text-generation/run_generation.py:
```diff
@@ -163,7 +178,7 @@ def main():
     set_seed(args.seed)
 
-    tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path)
+    tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path, add_prefix_space=True)
```
```suggestion
    if args.bad_words_ids is not None or args.force_words_ids is not None:
        tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path, add_prefix_space=True)
    else:
        tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path)
```
Let's add a prefix space only if needed for bad /forced words.

regisss commented on examples/text-generation/run_generation.py:
```diff
@@ -223,12 +238,22 @@ def main():
+    bad_words_ids = None
+    force_words_ids = None
+    if args.bad_words_ids is not None:
+        bad_words_ids = [tokenizer.encode(bad_word) for bad_word in args.bad_words_ids]
```
```suggestion
        bad_words_ids = [tokenizer.encode(bad_word, add_special_tokens=False) for bad_word in args.bad_words_ids]
```
As suggested here: https://huggingface.co/docs/transformers/main/en/internal/generation_utils#transformers.NoBadWordsLogitsProcessor

regisss commented on examples/text-generation/run_generation.py:
```diff
@@ -223,12 +238,22 @@ def main():
+    if args.bad_words_ids is not None:
+        bad_words_ids = [tokenizer.encode(bad_word) for bad_word in args.bad_words_ids]
+    if args.force_words_ids is not None:
+        force_words_ids = [tokenizer.encode(force_word) for force_word in args.force_words_ids]
```
```suggestion
        force_words_ids = [tokenizer.encode(force_word, add_special_tokens=False) for force_word in args.force_words_ids]
```

regisss commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -2271,4 +2271,179 @@ def constrained_beam_search(
+        beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
+        beam_scores[:, 1:] = -1e9
+        beam_scores = beam_scores.view((batch_size * num_beams,))
+        this_peer_finished = False  # used by synced_gpus only
```
```suggestion
        this_peer_finished = False  # used by synced_gpus only
        hb_profer = HabanaProfile(warmup=profiling_warmup_steps, active=profiling_steps)
        hb_profer.start()
```
Let's call the profiler if needed as for other decoding strategies

regisss commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -2271,4 +2271,179 @@ def constrained_beam_search(
+            if model_kwargs["past_key_values"] is not None:
+                model_kwargs["past_key_values"] = self._reorder_cache(model_kwargs["past_key_values"], beam_idx)
+            # increase cur_len
+            cur_len = cur_len + 1
```
```suggestion
            cur_len = cur_len + 1
            hb_profer.step()
```

regisss commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -2271,4 +2271,179 @@ def constrained_beam_search(
+                    break
+                else:
+                    this_peer_finished = True
+        sequence_outputs = constrained_beam_scorer.finalize(
```
```suggestion
        hb_profer.stop()
        sequence_outputs = constrained_beam_scorer.finalize(
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_281). All of your documentation changes will be reflected on that endpoint.

regisss reviewed (commented):
@vivekgoe Nice add to the library! I left several comments which should not take long to address I think.

I also see that the style check failed, could you run the following from the root of the repo please?
```bash
pip install black ruff
make style
```

I'm going to quickly test the PR and after all that it should be good to merge :slightly_smiling_face:

vivekgoe commented:
@regisss Thanks for your review comments. I have addressed issues highlighted and added a new commit to PR with the changes, please review this. 
Please note I have tested these configuration knobs for "gpt2" and "bigscience/bloom-560m".  All knobs work for gpt2, but for bloom-560m bad_words_ids does not work (same behavior is seen on CPU when using text-generation from transformers)

regisss reviewed (approved):
Looks good!

It sounds okay to me if BLOOM fails the same way as in Transformers.

I'm going to test it quickly with GPT2 and then I merge.

regisss commented:
@vivekgoe Could you share the command line you sued for GPT2?

I just tried the following:
```bash
python run_generation.py --model_name_or_path gpt2 --batch_size 1 --use_hpu_graphs --use_kv_cache --max_new_tokens 100 --bad_words_ids powerful
```
but it failed with a segmentation fault :thinking:

vivekgoe commented:
@regisss I can reproduce segmentation fault you reported, same command works if  "--num_beams 2" is added (which is how I tested).  Looks like something is going wrong when "bad_words_ids" is used with greedy_search.

regisss commented on examples/text-generation/run_generation.py:
```diff
@@ -114,6 +114,21 @@ def main():
         help="Optional argument to give a prompt of your choice as input.",
     )
+    parser.add_argument(
+        "--bad_words_ids",
```
I think we should rename this arg to `--bad_words` actually, since IDs are obtained after encoding with the tokenizer.

vivekgoe replied:
Done

regisss commented on examples/text-generation/run_generation.py:
```diff
@@ -114,6 +114,21 @@ def main():
+        help="Optional argument list of token ids that are not allowed to be generated.",
+    )
+    parser.add_argument(
+        "--force_words_ids",
```
Same with `--force_words`.

vivekgoe replied:
Done

regisss reviewed (commented):
Thanks @vivekgoe, it works now :slightly_smiling_face: 
I just left a couple of minor comments and then we'll merge!

vivekgoe commented:
@regisss uploaded commit with review comments addressed. Please help review and merge, thanks.

regisss reviewed (approved):
Let's merge it :rocket:

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_281). All of your documentation changes will be reflected on that endpoint.
